packages/api/src/upload.js

```diff
@@ -29,90 +11,33 @@ const INCREMENT_USER_USED_STORAGE = gql`
  */
 export async function uploadPost (request, env, ctx) {
-  const { user, authToken } = request.auth
   const { headers } = request
   const contentType = headers.get('content-type') || ''
 
-  let cid
-  let dagSize
   let name = headers.get('x-name')
-  let type
   if (!name || typeof name !== 'string') {
     name = `Upload at ${new Date().toISOString()}`
   }
 
+  let input
   if (contentType.includes('multipart/form-data')) {
     const form = await toFormData(request)
-    const files = /** @type {File[]} */ (form.getAll('file'))
-    const dirSize = files.reduce((total, f) => total + f.size, 0)
-
-    const entries = await env.cluster.addDirectory(files, {
-      metadata: { size: dirSize.toString() },
-      // When >2.5MB, use local add, because waiting for blocks to be sent to
-      // other cluster nodes can take a long time. Replication to other nodes
-      // will be done async by bitswap instead.
-      local: dirSize > LOCAL_ADD_THRESHOLD
-    })
-    const dir = entries[entries.length - 1]
-
-    cid = dir.cid
-    dagSize = dir.size
-    type = 'Multipart'
+    const files = form.getAll('file')
+    input = files.map((f) => ({
+      path: f.name,
+      content: f.stream()
+    }))
+  } else if (contentType.includes('application/car')) {
+    throw new Error('Please POST Content-addressed Archives to /car')
   } else {
     const blob = await request.blob()
     if (blob.size === 0) {
       throw new Error('Empty payload')
     }
-
-    const entry = await env.cluster.add(blob, {
-      metadata: { size: blob.size.toString() },
-      // When >2.5MB, use local add, because waiting for blocks to be sent to
-      // other cluster nodes can take a long time. Replication to other nodes
-      // will be done async by bitswap instead.
-      local: blob.size > LOCAL_ADD_THRESHOLD
-    })
-
-    cid = entry.cid
-    dagSize = entry.size
-    type = 'Blob'
-  }
-
-  // Retrieve current pin status and info about the nodes pinning the content.
-  const { peerMap } = await env.cluster.status(cid)
-  const pins = Object.entries(peerMap).map(([peerId, { peerName, status }]) => ({
-    status: toPinStatusEnum(status),
-    location: { peerId, peerName }
-  }))
-
-  if (!pins.length) { // should not happen
-    throw new Error('not pinning on any node')
-  }
-
-  // Store in DB
-  await env.db.query(CREATE_UPLOAD, {
-    data: {
-      user: user._id,
-      authToken: authToken?._id,
-      cid,
-      name,
-      type,
-      pins,
-      dagSize
-    }
-  })
-
-  if (ctx.waitUntil) {
-    ctx.waitUntil((async () => {
-      try {
-        await env.db.query(INCREMENT_USER_USED_STORAGE, {
-          user: user._id,
-          amount: dagSize
-        })
-      } catch (err) {
-        console.error(`failed to update user used storage: ${err.stack}`)
-      }
-    })())
+    input = [blob]
   }
-
-  return new JSONResponse({ cid })
+  // TODO: do we want to wrap file uploads like we do car uploads from the client?
```

Comment:
when receiving a single file, it won't have a path, while directories might have in the name. For directories with several files with no path like name it would likely make sense. I suggest that we should discuss this externally within an issue and keep the current behaviour in this PR context.